DNN Platform/DotNetNuke.ModulePipeline/ModuleControlPipeline.cs

```diff
@@ -0,0 +1,194 @@
+﻿using System;
+using System.IO;
+using DotNetNuke.Entities.Modules;
+using DotNetNuke.Instrumentation;
+using DotNetNuke.Services.Localization;
+using DotNetNuke.UI.Modules.Html5;
+using DotNetNuke.UI.Modules;
+using DotNetNuke.Web.Razor;
+using System.Collections.Generic;
+using DotNetNuke.Web.Mvc;
+
+#if NET472
+using System.Web.UI;
+#endif
+
+namespace DotNetNuke.ModulePipeline
+{
+    /// <summary>
+    /// The Module Pipeline that determines which Module pattern
+    /// to invoke based on the input module type.
+    /// </summary>
+    public class ModuleControlPipeline
+        // MULTI-TARGETTING PIPELINE
+        // -------------------------
+        // This file multi-targets .NET Framework and .NET Standard,
+        // which is needed as DNN migrates to .NET Core. The 'NET472'
+        // pre-processor directives are to fully support Legacy DNN.
+        // As the Pipeline is upgraded to be more complaint with 
+        // .NET Standard 2.0 use the apprioprate pre-processor directives.
+#if NET472
+        : IModuleControlPipeline
+#endif
+    {
+        private static readonly ILog TracelLogger = LoggerSource.Instance.GetLogger("DNN.Trace");
+        private Dictionary<string, IModuleControlFactory> _controlFactories;
+        public ModuleControlPipeline(
+            WebFormsModuleControlFactory webforms,
+            Html5ModuleControlFactory html5,
+            RazorModuleControlFactory razor3,
+            MvcModuleControlFactory mvc,
+            ReflectedModuleControlFactory fallthrough)
+        {
+            _controlFactories = new Dictionary<string, IModuleControlFactory>();
+            _controlFactories.Add(".ascx", webforms);
+            _controlFactories.Add(".htm", html5);
+            _controlFactories.Add(".html", html5);
+            _controlFactories.Add(".cshtml", razor3);
+            _controlFactories.Add(".vbhtml", razor3);
+            _controlFactories.Add(".mvc", mvc);
+            _controlFactories.Add("default", fallthrough);
+        }
+
+#if NET472
+        private IModuleControlFactory GetModuleControlFactory(string controlSrc)
+        {
+            string extension = Path.GetExtension(controlSrc.ToLowerInvariant());
+            IModuleControlFactory factory = _controlFactories[extension];
```

Comment:
This should use `TryGetValue` instead of the indexer, since a `Dictionary<T,U>` will throw when a key isn't found via the indexer.

Reply:
I was thinking we want the exception to be thrown to display the module error to the user that something is wrong with their code. There is a bigger try-catch statement around this code so the website won't crash but just the module code and display a DNN Error

Reply:
Don't we want it to use the fallback if the extension isn't in the dictionary?  My reading of the code is that it's valid (though uncommon) for an entry to be missing.

Reply:
You are ABSOLUTELY correct, I completely missed this. I will go ahead and fix it. Thanks for pointing this out.